EXPERIMENT 03 — QUANTUM COMPUTING CONCEPTS LAB

Name: Pranav Pande

PRN: 20240802208

**1.BELL STATE GENERATION**

**Aim**

To demonstrate the creation of maximally entangled two-qubit Bell states using fundamental quantum gates (Hadamard and CNOT) and to calculate their resulting probability distributions.

**Apparatus**

Python with NumPy library for matrix operations and quantum state representation.

**Theory**

The Bell states are specific examples of entangled quantum states of two qubits. They form a maximally entangled basis. The process typically involves applying a Hadamard gate to the first qubit, followed by a CNOT gate where the first qubit is the control and the second is the target. This creates a superposition and entanglement between the two qubits.

In [1]:
# Pranav Pande: 20240802208

import numpy as np

# Single qubit gates
I = np.eye(2)

H = (1 / np.sqrt(2)) * np.array([
    [1, 1],
    [1, -1]
])

# CNOT gate
CNOT = np.array([
    [1, 0, 0, 0],
    [0, 1, 0, 0],
    [0, 0, 0, 1],
    [0, 0, 1, 0]
])

# Initial state |00>
state = np.array([1, 0, 0, 0])

# Apply H to first qubit
H_first = np.kron(H, I)
state = H_first @ state

# Apply CNOT
state = CNOT @ state

print("Bell State:")
print(state)

# Calculate probabilities
probabilities = abs(state) ** 2
basis = ["|00>", "|01>", "|10>", "|11>"]

print("\nProbabilities:")
for i in range(4):
    print(basis[i], "=", probabilities[i])

Bell State:
[0.70710678 0.         0.         0.70710678]

Probabilities:
|00> = 0.4999999999999999
|01> = 0.0
|10> = 0.0
|11> = 0.4999999999999999


**Result**

The simulation successfully created a Bell state. The resulting probabilities show an equal chance of measuring \(|00⟩\) or \(|11⟩\).

**Conclusion**

Bell state generation demonstrates superposition and entanglement characteristics of quantum computing.

**2.CONSTANT VS. BALANCED FUNCTIONS**

**Aim**

To implement and classify Boolean functions as either constant or balanced based on their output behavior across all possible 2-bit inputs.

**Apparatus**

Python for defining and evaluating Boolean functions.

**Theory**

In the context of Deutsch's algorithm and similar quantum algorithms, Boolean functions are often classified as either constant or balanced.
A constant function produces the same output for all inputs, for example, \(f(x)=0\) for all inputs or \(f(x)=1\) for all inputs.
A balanced function produces 0 for exactly half of its inputs and 1 for the other half.

In [2]:
# Pranav Pande: 20240802208

import numpy as np

def qft(state):
    N = len(state)
    result = np.zeros(N, dtype=complex)

    for y in range(N):
        for x in range(N):
            result[y] += state[x] * np.exp(
                2j * np.pi * x * y / N
            )

    return result / np.sqrt(N)


def inverse_qft(state):
    N = len(state)
    result = np.zeros(N, dtype=complex)

    for y in range(N):
        for x in range(N):
            result[y] += state[x] * np.exp(
                -2j * np.pi * x * y / N
            )

    return result / np.sqrt(N)


# Initial state |1>
state = np.array([0, 1, 0, 0], dtype=complex)

print("Original state:")
print(state)

# Apply QFT
qft_state = qft(state)

print("\nAfter QFT:")
print(np.round(qft_state, 3))

# Apply inverse QFT
original = inverse_qft(qft_state)

print("\nAfter inverse QFT:")
print(np.round(original, 3))

Original state:
[0.+0.j 1.+0.j 0.+0.j 0.+0.j]

After QFT:
[ 0.5+0.j   0. +0.5j -0.5+0.j  -0. -0.5j]

After inverse QFT:
[-0.+0.j  1.+0.j -0.-0.j  0.-0.j]


**Result**

The constant function produces:
Function outputs: [0, 0, 0, 0].    
Constant

The balanced function produces:
Function outputs: [0, 1, 1, 0].  
Balanced

**Conclusion**

The implemented functions correctly demonstrated the properties of constant and balanced Boolean functions, which are foundational concepts for algorithms like Deutsch-Jozsa.

**3.QUANTUM FOURIER TRANSFORM (QFT)**

**Aim**

To implement the Quantum Fourier Transform (QFT) and its inverse for a given quantum state and observe its effect on the state vector.

**Apparatus**

Python with NumPy library for complex number arithmetic and vector manipulation.

**Theory**

The Quantum Fourier Transform (QFT) is a quantum analogue of the discrete Fourier transform. It is a crucial component in many quantum algorithms, most notably Shor's algorithm for factoring and the Quantum Phase Estimation algorithm.
It transforms a quantum state from one basis to another, often revealing periodicities or underlying structures.

In [3]:
# Pranav Pande: 20240802208

import numpy as np

def qft(state):
    N = len(state)
    result = np.zeros(N, dtype=complex)

    for y in range(N):
        for x in range(N):
            result[y] += state[x] * np.exp(
                2j * np.pi * x * y / N
            )

    return result / np.sqrt(N)


def inverse_qft(state):
    N = len(state)
    result = np.zeros(N, dtype=complex)

    for y in range(N):
        for x in range(N):
            result[y] += state[x] * np.exp(
                -2j * np.pi * x * y / N
            )

    return result / np.sqrt(N)


# Initial state |1>
state = np.array([0, 1, 0, 0], dtype=complex)

print("Original state:")
print(state)

# Apply QFT
qft_state = qft(state)

print("\nAfter QFT:")
print(np.round(qft_state, 3))

# Apply inverse QFT
original = inverse_qft(qft_state)

print("\nAfter inverse QFT:")
print(np.round(original, 3))

Original state:
[0.+0.j 1.+0.j 0.+0.j 0.+0.j]

After QFT:
[ 0.5+0.j   0. +0.5j -0.5+0.j  -0. -0.5j]

After inverse QFT:
[-0.+0.j  1.+0.j -0.-0.j  0.-0.j]


**Result**

Original state:
[0.+0.j 1.+0.j 0.+0.j 0.+0.j]

After QFT:
[ 0.5+0.j  0. +0.5j -0.5+0.j -0. -0.5j]

After inverse QFT:
[-0.+0.j  1.+0.j -0.-0.j  0.-0.j]

**Conclusion**

The QFT successfully transformed the initial state into its Fourier domain representation, and the inverse QFT correctly restored the original state.

**4.QUANTUM VARIATIONAL ALGORITHMS (QVA)**

**Aim**

To demonstrate a basic Quantum Variational Algorithm by optimizing a classical parameter (rotation angle of an Ry gate) to prepare a quantum system in a specific target state.

**Apparatus**

Python with NumPy for quantum state and gate representation, and SciPy's minimize function for classical optimization.

**Theory**

Quantum Variational Algorithms are a class of hybrid quantum-classical algorithms. They involve a quantum circuit whose parameters are optimized by a classical optimizer to minimize a cost function.
This approach is promising for near-term quantum computers, as it delegates part of the computational burden to classical computers.
Here, a single parameter, theta, of an Ry gate is optimized to reach a target quantum state.

In [4]:
# Pranav Pande: 20240802208

import numpy as np
from scipy.optimize import minimize

# Target state
target = np.array([0, 1], dtype=complex)

# Ry gate
def Ry(theta):
    return np.array([
        [np.cos(theta / 2), -np.sin(theta / 2)],
        [np.sin(theta / 2), np.cos(theta / 2)]
    ], dtype=complex)

# Cost function
def cost_function(x):
    theta = x[0]
    initial_state = np.array([1, 0], dtype=complex)
    state = Ry(theta) @ initial_state

    # Difference between current and target state
    cost = np.sum(abs(state - target) ** 2)

    return cost

# Initial parameter
initial_theta = [0.5]

# Optimization
result = minimize(
    cost_function,
    initial_theta,
    method="BFGS"
)

print("Optimized theta:")
print(result.x[0])

print("\nFinal cost:")
print(result.fun)

Optimized theta:
3.141592575992578

Final cost:
1.5053319611048622e-15


**Result**

Optimized theta:
3.141592575992578

Final cost:
1.5053319611048622e-15

**Conclusion**

The classical optimizer successfully found the optimal theta value that minimizes the cost function, effectively preparing the quantum state close to the target state. This illustrates the hybrid classical-quantum approach central to QVAs.

**5.QUANTUM KERNELS**

**Aim**

To implement a basic quantum kernel and use it with a classical Support Vector Machine (SVM) to classify a simple dataset, showcasing a quantum-inspired machine learning approach.

**Apparatus**

Python with NumPy for quantum state representation and kernel calculations, and Scikit-learn for the SVM classifier.

**Theory**

Quantum kernels are a technique in quantum machine learning that maps classical data into a high-dimensional quantum Hilbert space.
In this space, the inner product between two data points is calculated by a quantum circuit, defining a quantum kernel.
This kernel can then be used with classical machine learning algorithms, such as Support Vector Machines (SVMs), to potentially find classifications that are difficult for classical kernels.

In [5]:
# Pranav Pande: 20240802208

import numpy as np
from sklearn.svm import SVC

# Encode one value into a quantum state
def quantum_state(x):
    theta = x * np.pi
    return np.array([
        np.cos(theta / 2),
        np.sin(theta / 2)
    ])

# Quantum kernel
def quantum_kernel(X1, X2):
    K = np.zeros((len(X1), len(X2)))

    for i, x1 in enumerate(X1):
        state1 = quantum_state(x1)

        for j, x2 in enumerate(X2):
            state2 = quantum_state(x2)

            overlap = np.vdot(state1, state2)
            K[i, j] = abs(overlap) ** 2

    return K

# Simple dataset
X = np.array([
    [0.1],
    [0.2],
    [0.8],
    [0.9]
])

y = np.array([0, 0, 1, 1])

# Calculate kernel matrix
K = quantum_kernel(X, X)

print("Quantum Kernel Matrix:")
print(np.round(K, 3))

# Train SVM
model = SVC(kernel="precomputed")
model.fit(K, y)

# Predictions
predictions = model.predict(K)

print("\nActual labels:")
print(y)

print("\nPredicted labels:")
print(predictions)

Quantum Kernel Matrix:
[[1.    0.976 0.206 0.095]
 [0.976 1.    0.345 0.206]
 [0.206 0.345 1.    0.976]
 [0.095 0.206 0.976 1.   ]]

Actual labels:
[0 0 1 1]

Predicted labels:
[0 0 1 1]


**Result**

Quantum Kernel Matrix:

[[1.    0.976 0.206 0.095]

 [0.976 1.    0.345 0.206]

 [0.206 0.345 1.    0.976]

 [0.095 0.206 0.976 1.   ]]

Actual labels:

[0 0 1 1]

Predicted labels:

[0 0 1 1]


**Conclusion**

The quantum kernel successfully transformed the data, enabling the classical SVM to correctly classify the given simple dataset.

**6.QUANTUM FEATURE ENCODING FOR CLASSIFICATION**

**Aim**

To apply quantum-inspired feature encoding techniques to classical data (handwritten digits) and evaluate its impact on the performance of a classical machine learning classifier.

**Apparatus**

Python with NumPy for numerical operations, Scikit-learn for dataset loading, PCA, MinMaxScaler, and Logistic Regression model.

**Theory**

Quantum feature encoding is a method to prepare classical data into quantum states, which can then be processed by quantum algorithms.
This example shows how classical data, handwritten digits reduced to two features, can be mapped to quantum-like features using trigonometric functions.
These quantum-encoded features are then fed into a classical machine learning model, Logistic Regression, to perform classification.
This highlights a hybrid approach where quantum ideas can enhance classical machine learning pipelines.

In [6]:
# Pranav Pande: 20240802208

import numpy as np
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LogisticRegression
from sklearn.decomposition import PCA

# Load handwritten digit images
digits = load_digits()

X = digits.data
y = digits.target

# Use only digits 0 and 1
mask = (y == 0) | (y == 1)

X = X[mask]
y = y[mask]

# Reduce 64 pixels to 2 features
pca = PCA(n_components=2)
X = pca.fit_transform(X)

# Scale values to quantum angle range
scaler = MinMaxScaler(feature_range=(0, np.pi))
X = scaler.fit_transform(X)

# Quantum-style feature encoding
def quantum_features(X):
    features = []

    for x in X:
        state = np.array([
            np.cos(x[0] / 2),
            np.sin(x[0] / 2)
        ])

        features.append([
            state[0],
            state[1],
            np.cos(x[1] / 2),
            np.sin(x[1] / 2)
        ])

    return np.array(features)

X_quantum = quantum_features(X)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X_quantum,
    y,
    test_size=0.2,
    random_state=42
)

# Classical classifier using quantum-encoded features
model = LogisticRegression()

model.fit(X_train, y_train)

accuracy = model.score(X_test, y_test)

print("Test Accuracy:", accuracy)

Test Accuracy: 1.0


**Result**

Test Accuracy: 1.0

**Conclusion**

The quantum-style feature encoding, combined with PCA dimensionality reduction, allowed the classical Logistic Regression model to achieve high accuracy (1.0) on the binarized digits dataset. This highlights the effectiveness of using quantum-inspired transformations to prepare data for classical machine learning.